documentation/ADRs/020_Log_Files_and_Real-Time_Alerts.md

```diff
@@ -0,0 +1,75 @@
+
+| ADR Info            | Details                   |
+|---------------------|---------------------------|
+| Subject             | Log files and real-time alerts |
+| ADR Number          | 020                       |
+| Status              | Proposed                  |
+| Author              | Marina                    |
+| Date                | 28.10.2024                |
+
+
+## Context
+In order to have a more effective overview of the orchestration in the pipeline, there is a need for better handling of log files and implementing real-time alerts which would allow for timely and relevant error handling for critical issues. 
+
+
+## Decision
+To implement a real-time alert and notification system which will distribute alerts through specific channels (eg. Slack, email, W&B, Prefect), targeted at particular audiences (eg. Infrastructure, MD&D, individual) depending on the type of alert being handled. 
+
+
+### Overview
+
+Alert channels: Slack, email, Prefect and Weights&Biases
+
+Logging levels: INFO, WARNING, ERROR, CRITICAL
+
+Audience: MD&D, Infrastructure, Outreach, monthly-run responsible 
+
+| Channel | Threshold        | Alert Type                         | Audience                  |
+|---------|-------------------|------------------------------------|---------------------------|
+| Slack   | ERROR, CRITICAL   | Real-Time, Short Summary + Link to Full Log | MD&D, Infrastructure      |
+| Email   | ERROR, CRITICAL   | Real-Time, Expanded Summary       | MD&D, Infrastructure      |
+| Prefect | ALL               | Real-Time, Workflow Status        | MD&D, Infrastructure, Outreach |
+| W&B     | ALL               | Performance Metrics               | MD&D, Infrastructure, Outreach |
+
+
+For alert prioritization, "critical" and "non-critical" alerts are separated, with critical alerts stemming from logging levels ERROR and CRITICAL. 
+
+
+## Consequences
+
+**Positive Effects:**
+- Real-time alerts allow for timely handling of errors which may be critical 
+- Alerts can reach their target audiences immediately which limits time and makes information flows more direct and effective 
+- Having real-time alerts is useful in generating a record of (reoccuring) issues in the pipeline 
+- Increased transparency
+
+
+**Negative Effects:**
+- Alerts must be set up in a way to be concise and informative in order to be useful, with carefully chosen levels of detail to avoid "alert fatigue"
+- The audience for each type of alert must be carefully selected in order for alerts to stay relevant 
+- Potential increase in maintanence work 
+
+## Rationale
+Easier, quicker and more efficient handling of potential issues, errors or obstacles in the pipeline. Additionally, a better overview of the pipeline orchestration process. 
+
+### Considerations
+
+Implementation: 
+- how the channels are integrated with the logger
+- what types of events or errors trigger alerts at each logging level 
+- testing accuracy and reliability of the alerts
```

Comment:
I would consider being more detailed here as well:

#### Implementation

- **Logger Integration:**
  - Choose a compatible logging framework: we'll use Python’s `logging`.
  - Connect the logger to alert channels using middleware (e.i., Slack API, Mail).
  - Integrate with Prefect and Weights & Biases (W&B) for workflow and performance metrics.

- **Event and Error Triggers:**
  - Define criteria and examples for each logging level (INFO, WARNING, ERROR, CRITICAL).
  - Use structured logging (e.g., JSON) for easy log parsing and filtering.
  - Set rules for triggering alerts based on event severity.

- **Testing and Reliability:**
  - Implement a testing strategy with unit, integration, and end-to-end tests for alert verification.
  - Simulate error conditions to validate alert accuracy and reliability.
  - Monitor the alerting system to detect and resolve delivery issues.